src/controller.ts

```diff
@@ -200,20 +199,23 @@ export class Controller {
       throw new Error("Invalid compilation result");
     }
 
-    let mostDeepSigCell = cell.refs[1];
-
-    while (true) {
-      if (mostDeepSigCell.refs.length === 0) {
-        break;
-      }
-      mostDeepSigCell = mostDeepSigCell.refs[0];
-    }
-
-    const { sigCell } = signatureCell(cell.refs[0], this.keypair);
-    mostDeepSigCell.refs.push(sigCell);
+    const slice = cell.beginParse();
+    const msgToSign = slice.loadRef();
+    const { sigCell } = signatureCell(msgToSign, this.keypair);
+    let child = storeSign(slice.loadRef(), sigCell);
 
     return {
-      msgCell: cell.toBoc(),
+      msgCell: slice.asBuilder().storeRef(msgToSign).storeRef(child).asCell().toBoc(),
     };
+
+    function storeSign(node: Cell, sigCell: Cell): Cell {
+      const slice = node.beginParse();
+      if (slice.remainingRefs > 0) {
+        const child = storeSign(slice.loadRef(), msgToSign);
+        return slice.asBuilder().storeRef(child).asCell();
+      } else {
+        return slice.asBuilder().storeRef(sigCell).asCell();
+      }
+    }
```

Comment:
I think this could make more sense maybe as a recursive pattern (each path has one job, either to find the deepest cell or to store sigCell within it)

Reply:
The previous implementation (below) used a loop to mutate the `cell` object and put the signed message at the deepest level. So I added this recursive function that finds the deepest cell, stores the signed message and then going up, reconstructs the object. So your suggestion does not do the reconstruction part.

```python
let mostDeepSigCell = cell.refs[1];

while (true) {
  if (mostDeepSigCell.refs.length === 0) {
    break;
  }
  mostDeepSigCell = mostDeepSigCell.refs[0];
}

const { sigCell } = signatureCell(cell.refs[0], this.keypair);
mostDeepSigCell.refs.push(sigCell);

return {
  msgCell: cell.toBoc(),
};
```

Reply:
So this is because asBuilder() breaks the reference. Can we still use cell.refs.push? that'd enable either the loop or a simpler recursive function.

Otherwise I think suggested recursive implementation would cause this:

If we have a `sigCell1(c1)->sigCell2(c2)` and we're trying to append a new `sigCell3(c3)`, then our recursive call tree is:
```
// assuming we're sending sigCell and not msgToSign, since we only want to store sigCell at the end
storeSign(c1, c3)
	storeSign(c2, c3)
           // this will create a new cell with the contents of c2, with a ref to c3
        // this new cell (c2->c3) is returned here and is store onto a new cell, again with the contents of c2, now pointing to what was returned from the previous recursive call, such that c2->c2->c3, which is not correct.
```